File: compat/suites/go-sdk/internal/groups/cognito.go
package groups

import (
	"context"
	"fmt"

	"github.com/Neaox/overcast-compat-go-sdk/internal/clients"
	"github.com/Neaox/overcast-compat-go-sdk/internal/harness"
	"github.com/aws/aws-sdk-go-v2/aws"
	cip "github.com/aws/aws-sdk-go-v2/service/cognitoidentityprovider"
	ciptypes "github.com/aws/aws-sdk-go-v2/service/cognitoidentityprovider/types"
)

func Cognito(c *clients.Clients) ServiceGroup {
	g := &cognitoGroup{c: c}
	return ServiceGroup{
		Impls: map[string]harness.TestFn{
			"CreateUserPool":       g.CreateUserPool,
			"DescribeUserPool":     g.DescribeUserPool,
			"ListUserPools":        g.ListUserPools,
			"CreateUserPoolClient": g.CreateUserPoolClient,
			"ListUserPoolClients":  g.ListUserPoolClients,
			"AdminCreateUser":      g.AdminCreateUser,
			"ListUsers":            g.ListUsers,
			"AdminDeleteUser":      g.AdminDeleteUser,
			"DeleteUserPool":       g.DeleteUserPool,
			"CreateUserPoolClient with token validity":      g.CreateClientTokenValidity,
			"DescribeUserPoolClient returns token validity": g.DescribeClientTokenValidity,
			"UpdateUserPoolClient changes token validity":   g.UpdateClientTokenValidity,
			"DeleteUserPoolClient":                          g.DeleteUserPoolClient,
		},
		Setup: map[string]func(context.Context, *harness.TestContext) error{
			"cognito-userpools":      g.setupUserPools,
			"cognito-token-validity": g.setupTokenValidity,
		},
		Teardown: map[string]func(context.Context, *harness.TestContext) error{
			"cognito-userpools":      g.teardownUserPools,
			"cognito-token-validity": g.teardownTokenValidity,
		},
	}
}

type cognitoGroup struct{ c *clients.Clients }

func (g *cognitoGroup) cl() *cip.Client { return g.c.Cognito() }

func (g *cognitoGroup) setupUserPools(_ context.Context, _ *harness.TestContext) error { return nil }

func (g *cognitoGroup) teardownUserPools(ctx context.Context, t *harness.TestContext) error {
	poolID := t.GetString("cognito_pool_id")
	if poolID == "" {
		return nil
	}
	if username := t.GetString("cognito_username"); username != "" {
		g.cl().AdminDeleteUser(ctx, &cip.AdminDeleteUserInput{ //nolint:errcheck
			UserPoolId: aws.String(poolID),
			Username:   aws.String(username),
		})
	}
	g.cl().DeleteUserPool(ctx, &cip.DeleteUserPoolInput{UserPoolId: aws.String(poolID)}) //nolint:errcheck
	return nil
}

func (g *cognitoGroup) CreateUserPool(ctx context.Context, t *harness.TestContext) error {
	poolName := fmt.Sprintf("compat-%s", t.RunID)
	resp, err := g.cl().CreateUserPool(ctx, &cip.CreateUserPoolInput{
		PoolName: aws.String(poolName),
	})
	if err != nil {
		return err
	}
	if resp.UserPool == nil || resp.UserPool.Id == nil {
		return fmt.Errorf("CreateUserPool: missing Id")
	}
	t.Set("cognito_pool_id", *resp.UserPool.Id)
	return nil
}

func (g *cognitoGroup) DescribeUserPool(ctx context.Context, t *harness.TestContext) error {
	poolID := t.GetString("cognito_pool_id")
	if poolID == "" {
		return fmt.Errorf("DescribeUserPool: no pool from CreateUserPool")
	}
	resp, err := g.cl().DescribeUserPool(ctx, &cip.DescribeUserPoolInput{
		UserPoolId: aws.String(poolID),
	})
	if err != nil {
		return err
	}
	if resp.UserPool == nil || resp.UserPool.Id == nil {
		return fmt.Errorf("DescribeUserPool: missing Id")
	}
	return nil
}

func (g *cognitoGroup) ListUserPools(ctx context.Context, t *harness.TestContext) error {
	_, err := g.cl().ListUserPools(ctx, &cip.ListUserPoolsInput{
		MaxResults: aws.Int32(10),
	})
	return err
}

func (g *cognitoGroup) AdminCreateUser(ctx context.Context, t *harness.TestContext) error {
	poolID := t.GetString("cognito_pool_id")
	if poolID == "" {
		return fmt.Errorf("AdminCreateUser: no pool from CreateUserPool")
	}
	username := fmt.Sprintf("compat-user-%s", t.RunID)
	_, err := g.cl().AdminCreateUser(ctx, &cip.AdminCreateUserInput{
		UserPoolId: aws.String(poolID),
		Username:   aws.String(username),
	})
	if err != nil {
		return err
	}
	t.Set("cognito_username", username)
	return nil
}

func (g *cognitoGroup) ListUsers(ctx context.Context, t *harness.TestContext) error {
	poolID := t.GetString("cognito_pool_id")
	if poolID == "" {
		return nil
	}
	_, err := g.cl().ListUsers(ctx, &cip.ListUsersInput{
		UserPoolId: aws.String(poolID),
	})
	return err
}

func (g *cognitoGroup) AdminDeleteUser(ctx context.Context, t *harness.TestContext) error {
	poolID := t.GetString("cognito_pool_id")
	username := t.GetString("cognito_username")
	if poolID == "" || username == "" {
		return nil
	}
	_, err := g.cl().AdminDeleteUser(ctx, &cip.AdminDeleteUserInput{
		UserPoolId: aws.String(poolID),
		Username:   aws.String(username),
	})
	return err
}

func (g *cognitoGroup) DeleteUserPool(ctx context.Context, t *harness.TestContext) error {
	poolID := t.GetString("cognito_pool_id")
	if poolID == "" {
		return nil
	}
	_, err := g.cl().DeleteUserPool(ctx, &cip.DeleteUserPoolInput{
		UserPoolId: aws.String(poolID),
	})
	return err
}

func (g *cognitoGroup) CreateUserPoolClient(ctx context.Context, t *harness.TestContext) error {
	poolID := t.GetString("cognito_pool_id")
	if poolID == "" {
		return fmt.Errorf("CreateUserPoolClient: no pool from CreateUserPool")
	}
	resp, err := g.cl().CreateUserPoolClient(ctx, &cip.CreateUserPoolClientInput{
		UserPoolId: aws.String(poolID),
		ClientName: aws.String(fmt.Sprintf("compat-client-%s", t.RunID)),
	})
	if err != nil {
		return err
	}
	if resp.UserPoolClient == nil || resp.UserPoolClient.ClientId == nil {
		return fmt.Errorf("CreateUserPoolClient: missing ClientId")
	}
	t.Set("cognito_client_id", *resp.UserPoolClient.ClientId)
	return nil
}

func (g *cognitoGroup) ListUserPoolClients(ctx context.Context, t *harness.TestContext) error {
	poolID := t.GetString("cognito_pool_id")
	if poolID == "" {
		return fmt.Errorf("ListUserPoolClients: no pool from CreateUserPool")
	}
	_, err := g.cl().ListUserPoolClients(ctx, &cip.ListUserPoolClientsInput{
		UserPoolId: aws.String(poolID),
		MaxResults: aws.Int32(10),
	})
	return err
}

// ── cognito-token-validity ─────────────────────────────────────────────────

func (g *cognitoGroup) setupTokenValidity(_ context.Context, _ *harness.TestContext) error {
	return nil
}

func (g *cognitoGroup) teardownTokenValidity(ctx context.Context, t *harness.TestContext) error {
	poolID := t.GetString("tv_pool_id")
	clientID := t.GetString("tv_client_id")
	if poolID != "" && clientID != "" {
		g.cl().DeleteUserPoolClient(ctx, &cip.DeleteUserPoolClientInput{ //nolint:errcheck
			UserPoolId: aws.String(poolID),
			ClientId:   aws.String(clientID),
		})
	}
	if poolID != "" {
		g.cl().DeleteUserPool(ctx, &cip.DeleteUserPoolInput{UserPoolId: aws.String(poolID)}) //nolint:errcheck
	}
	return nil
}

func (g *cognitoGroup) CreateClientTokenValidity(ctx context.Context, t *harness.TestContext) error {
	poolName := fmt.Sprintf("compat-tv-%s", t.RunID)
	poolResp, err := g.cl().CreateUserPool(ctx, &cip.CreateUserPoolInput{
		PoolName: aws.String(poolName),
	})
	if err != nil {
		return err
	}
	poolID := *poolResp.UserPool.Id
	t.Set("tv_pool_id", poolID)

	resp, err := g.cl().CreateUserPoolClient(ctx, &cip.CreateUserPoolClientInput{
		UserPoolId:           aws.String(poolID),
		ClientName:           aws.String(fmt.Sprintf("compat-client-%s", t.RunID)),
		AccessTokenValidity:  aws.Int32(2),
		IdTokenValidity:      aws.Int32(3),
		RefreshTokenValidity: 7,
		TokenValidityUnits: &ciptypes.TokenValidityUnitsType{
			AccessToken:  ciptypes.TimeUnitsTypeHours,
			IdToken:      ciptypes.TimeUnitsTypeHours,
			RefreshToken: ciptypes.TimeUnitsTypeDays,
		},
	})
	if err != nil {
		return err
	}
	if resp.UserPoolClient == nil || resp.UserPoolClient.ClientId == nil {
		return fmt.Errorf("CreateClientTokenValidity: missing ClientId")
	}
	t.Set("tv_client_id", *resp.UserPoolClient.ClientId)
	return nil
}

func (g *cognitoGroup) DescribeClientTokenValidity(ctx context.Context, t *harness.TestContext) error {
	poolID := t.GetString("tv_pool_id")
	clientID := t.GetString("tv_client_id")
	if poolID == "" || clientID == "" {
		return fmt.Errorf("DescribeClientTokenValidity: missing pool/client from create")
	}
	_, err := g.cl().DescribeUserPoolClient(ctx, &cip.DescribeUserPoolClientInput{
		UserPoolId: aws.String(poolID),
		ClientId:   aws.String(clientID),
	})
	return err
}

func (g *cognitoGroup) UpdateClientTokenValidity(ctx context.Context, t *harness.TestContext) error {
	poolID := t.GetString("tv_pool_id")
	clientID := t.GetString("tv_client_id")
	if poolID == "" || clientID == "" {
		return fmt.Errorf("UpdateClientTokenValidity: missing pool/client from create")
	}
	_, err := g.cl().UpdateUserPoolClient(ctx, &cip.UpdateUserPoolClientInput{
		UserPoolId:          aws.String(poolID),
		ClientId:            aws.String(clientID),
		AccessTokenValidity: aws.Int32(30),
		TokenValidityUnits: &ciptypes.TokenValidityUnitsType{
			AccessToken:  ciptypes.TimeUnitsTypeMinutes,
			IdToken:      ciptypes.TimeUnitsTypeHours,
			RefreshToken: ciptypes.TimeUnitsTypeDays,
		},
	})
	return err
}

func (g *cognitoGroup) DeleteUserPoolClient(ctx context.Context, t *harness.TestContext) error {
	poolID := t.GetString("tv_pool_id")
	clientID := t.GetString("tv_client_id")
	if poolID == "" || clientID == "" {
		return nil
	}
	_, err := g.cl().DeleteUserPoolClient(ctx, &cip.DeleteUserPoolClientInput{
		UserPoolId: aws.String(poolID),
		ClientId:   aws.String(clientID),
	})
	return err
}

File: compat/suites/python-sdk/groups/cognito.py
"""
groups/cognito.py — Cognito User Pools compatibility test implementations for the Python suite.
"""

from __future__ import annotations
from lib.harness import TestContext
from lib.clients import make_clients


def _cognito(ctx: TestContext):
    return make_clients(ctx.endpoint, ctx.region)._get("cognito-idp")


# ── cognito-userpools ─────────────────────────────────────────────────────────

def CreateUserPool(ctx: TestContext) -> None:
    cog = _cognito(ctx)
    pool_name = f"compat-{ctx.run_id}"
    resp = cog.create_user_pool(PoolName=pool_name)
    pool = resp.get("UserPool", {})
    if not pool.get("Id"):
        raise AssertionError("CreateUserPool: missing Id")
    ctx["cognito_pool_id"] = pool["Id"]


def DescribeUserPool(ctx: TestContext) -> None:
    cog = _cognito(ctx)
    pool_id = ctx.get("cognito_pool_id")
    if not pool_id:
        raise AssertionError("DescribeUserPool: no pool from CreateUserPool")
    resp = cog.describe_user_pool(UserPoolId=pool_id)
    if not resp.get("UserPool", {}).get("Id"):
        raise AssertionError("DescribeUserPool: missing Id")


def ListUserPools(ctx: TestContext) -> None:
    _cognito(ctx).list_user_pools(MaxResults=10)


def AdminCreateUser(ctx: TestContext) -> None:
    cog = _cognito(ctx)
    pool_id = ctx.get("cognito_pool_id")
    if not pool_id:
        raise AssertionError("AdminCreateUser: no pool from CreateUserPool")
    username = f"compat-user-{ctx.run_id}"
    cog.admin_create_user(UserPoolId=pool_id, Username=username)
    ctx["cognito_username"] = username


def ListUsers(ctx: TestContext) -> None:
    cog = _cognito(ctx)
    pool_id = ctx.get("cognito_pool_id")
    if not pool_id:
        return
    cog.list_users(UserPoolId=pool_id)


def AdminDeleteUser(ctx: TestContext) -> None:
    cog = _cognito(ctx)
    pool_id = ctx.get("cognito_pool_id")
    username = ctx.get("cognito_username")
    if not pool_id or not username:
        return
    cog.admin_delete_user(UserPoolId=pool_id, Username=username)


def DeleteUserPool(ctx: TestContext) -> None:
    cog = _cognito(ctx)
    pool_id = ctx.get("cognito_pool_id")
    if not pool_id:
        return
    cog.delete_user_pool(UserPoolId=pool_id)


# ── ImplMap ───────────────────────────────────────────────────────────────────

IMPLS = {
    "CreateUserPool": CreateUserPool,
    "DescribeUserPool": DescribeUserPool,
    "ListUserPools": ListUserPools,
    "CreateUserPoolClient": lambda ctx: CreateUserPoolClientFn(ctx),
    "ListUserPoolClients": lambda ctx: ListUserPoolClientsFn(ctx),
    "AdminCreateUser": AdminCreateUser,
    "ListUsers": ListUsers,
    "AdminDeleteUser": AdminDeleteUser,
    "DeleteUserPool": DeleteUserPool,
    "CreateUserPoolClient with token validity": lambda ctx: CreateClientTokenValidity(ctx),
    "DescribeUserPoolClient returns token validity": lambda ctx: DescribeClientTokenValidity(ctx),
    "UpdateUserPoolClient changes token validity": lambda ctx: UpdateClientTokenValidity(ctx),
    "DeleteUserPoolClient": lambda ctx: DeleteUserPoolClientFn(ctx),
}

SETUP = {
    "cognito-token-validity": lambda ctx: None,
}
TEARDOWN = {
    "cognito-userpools": lambda ctx: _teardown_user_pool(ctx),
    "cognito-token-validity": lambda ctx: _teardown_token_validity(ctx),
}


def _teardown_user_pool(ctx: TestContext) -> None:
    pool_id = ctx.get("cognito_pool_id")
    if not pool_id:
        return
    username = ctx.get("cognito_username")
    if username:
        try:
            _cognito(ctx).admin_delete_user(UserPoolId=pool_id, Username=username)
        except Exception:
            pass
    try:
        _cognito(ctx).delete_user_pool(UserPoolId=pool_id)
    except Exception:
        pass


def CreateUserPoolClientFn(ctx: TestContext) -> None:
    cog = _cognito(ctx)
    pool_id = ctx.get("cognito_pool_id")
    if not pool_id:
        raise AssertionError("CreateUserPoolClient: no pool from CreateUserPool")
    resp = cog.create_user_pool_client(
        UserPoolId=pool_id,
        ClientName=f"compat-client-{ctx.run_id}",
    )
    client = resp.get("UserPoolClient", {})
    if not client.get("ClientId"):
        raise AssertionError("CreateUserPoolClient: missing ClientId")
    ctx["cognito_client_id"] = client["ClientId"]


def ListUserPoolClientsFn(ctx: TestContext) -> None:
    cog = _cognito(ctx)
    pool_id = ctx.get("cognito_pool_id")
    if not pool_id:
        raise AssertionError("ListUserPoolClients: no pool from CreateUserPool")
    cog.list_user_pool_clients(UserPoolId=pool_id, MaxResults=10)


# ── cognito-token-validity ────────────────────────────────────────────────────

def _teardown_token_validity(ctx: TestContext) -> None:
    cog = _cognito(ctx)
    pool_id = ctx.get("tv_pool_id")
    client_id = ctx.get("tv_client_id")
    if pool_id and client_id:
        try:
            cog.delete_user_pool_client(UserPoolId=pool_id, ClientId=client_id)
        except Exception:
            pass
    if pool_id:
        try:
            cog.delete_user_pool(UserPoolId=pool_id)
        except Exception:
            pass


def CreateClientTokenValidity(ctx: TestContext) -> None:
    cog = _cognito(ctx)
    pool_name = f"compat-tv-{ctx.run_id}"
    resp = cog.create_user_pool(PoolName=pool_name)
    pool_id = resp.get("UserPool", {}).get("Id")
    if not pool_id:
        raise AssertionError("CreateClientTokenValidity: missing pool Id")
    ctx["tv_pool_id"] = pool_id

    resp = cog.create_user_pool_client(
        UserPoolId=pool_id,
        ClientName=f"compat-client-{ctx.run_id}",
        AccessTokenValidity=2,
        IdTokenValidity=3,
        RefreshTokenValidity=7,
        TokenValidityUnits={
            "AccessToken": "hours",
            "IdToken": "hours",
            "RefreshToken": "days",
        },
    )
    client = resp.get("UserPoolClient", {})
    if not client.get("ClientId"):
        raise AssertionError("CreateClientTokenValidity: missing ClientId")
    ctx["tv_client_id"] = client["ClientId"]


def DescribeClientTokenValidity(ctx: TestContext) -> None:
    cog = _cognito(ctx)
    pool_id = ctx.get("tv_pool_id")
    client_id = ctx.get("tv_client_id")
    if not pool_id or not client_id:
        raise AssertionError("DescribeClientTokenValidity: missing pool/client")
    cog.describe_user_pool_client(UserPoolId=pool_id, ClientId=client_id)


def UpdateClientTokenValidity(ctx: TestContext) -> None:
    cog = _cognito(ctx)
    pool_id = ctx.get("tv_pool_id")
    client_id = ctx.get("tv_client_id")
    if not pool_id or not client_id:
        raise AssertionError("UpdateClientTokenValidity: missing pool/client")
    cog.update_user_pool_client(
        UserPoolId=pool_id,
        ClientId=client_id,
        AccessTokenValidity=30,
        TokenValidityUnits={
            "AccessToken": "minutes",
            "IdToken": "hours",
            "RefreshToken": "days",
        },
    )


def DeleteUserPoolClientFn(ctx: TestContext) -> None:
    cog = _cognito(ctx)
    pool_id = ctx.get("tv_pool_id")
    client_id = ctx.get("tv_client_id")
    if not pool_id or not client_id:
        return
    cog.delete_user_pool_client(UserPoolId=pool_id, ClientId=client_id)
